### PR TITLE
Fix invoke_subgraph eager autograd execution

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #184700

Run invoke_subgraph directly under real-tensor eager execution so normal autograd records through the subgraph instead of entering the fake-tensor AOTAutograd backward path.

Fixes #178520

Generated by my agent

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98 @mlazos

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

File: test/dynamo/test_modes.py [hunk 1]
```diff
@@ -931,6 +931,64 @@ def __torch_function__(self, *args, **kwargs):
...
+        def train_step(inp):
+            return run_step(model, inp)
+
+        backend = torch._dynamo.testing.EagerAndRecordGraphs()
+        compiled = torch.compile(train_step, backend=backend, fullgraph=True)
+        loss = compiled(x.clone())
+
+        self.assertEqual(loss, ref_loss)
+        self.assertEqual(len(backend.graphs), 1)
+        self.assertIn("invoke_subgraph", backend.graphs[0].code)
```
Comment (aorenste) [inline test/dynamo/test_modes.py]: (Claude Review) [cleanup] `assertIn("invoke_subgraph", backend.graphs[0].code)` is a loose substring check. It would also match helper names that happen to contain the substring, and the printed FX `code` text is brittle to cosmetic graph changes. Other tests in this class use `assertExpectedInline(normalize_gm(backend.graphs[0]))` (e.g. lines 1027-1029, 1079-1084) which locks down the exact placement of `torch.ops.higher_order.invoke_subgraph` and would catch a regression where the captured graph silently inlines the region instead of emitting the HOP. Alternatively, scanning `backend.graphs[0].graph.nodes` for a node whose `target` is `torch.ops.higher_order.invoke_subgraph` is a tighter structural check. The numerical loss/gradient comparison above is the load-bearing assertion; this one is largely a smoke check today.

File: torch/_higher_order_ops/invoke_subgraph.py [hunk 2]
```diff
@@ -923,6 +923,15 @@ def backward(
 
 @invoke_subgraph.py_autograd_impl
 def _(subgraph, identifier, *operands):
+    from torch._guards import detect_fake_mode
+
+    # Eager backends run the captured HOP with real tensors and no AOTAutograd
+    # fake mode. Let regular autograd record through the subgraph in that case.
+    if detect_fake_mode(operands) is None:
+        if getattr(subgraph, "_boxed_call", False):
+            return subgraph(list(operands))
```
Comment (aorenste) [inline torch/_higher_order_ops/invoke_subgraph.py:960]: Not this PR but we really should have a helper for this pattern...

**Review by aorenste — APPROVED** [review]

File: torch/_higher_order_ops/invoke_subgraph.py (hunk 2 above)
Comment (jansel) [reply torch/_higher_order_ops/invoke_subgraph.py:960]: My agent says I agree that a helper would be useful, and I kept that as a separate cleanup since this PR is scoped to the eager-autograd fix. The review-driven update only tightens the regression assertion.

**Comment by jansel** [comment]
My agent says I rebased and resubmitted this PR onto current main after mergebot timed out on the cancelled trunk / before-test / llm-retrieval job. The focused regression test and lintrunner -a both pass locally; CI has restarted on the new head.

**Comment by jansel** [comment]
My agent says I rebased and resubmitted this PR onto current main after the remaining docs jobs failed during preview upload with `aws: command not found`, which appears to be a workflow/runner environment issue after docs generation completed. The focused regression test and `lintrunner -a` both pass locally; CI has restarted on the refreshed head.

**Comment by jansel** [comment]
My agent says the remaining inductor HuggingFace failure was a DebertaV2ForMaskedLM eager import/JIT scripting failure (`aten::manual_seed` schema mismatch) in the benchmark harness, with a separate gemma OOM logged later; neither touches this PR

**Comment by jansel** [comment]
My agent says I rebased and resubmitted onto current main after that unrelated benchmark-harness failure. The focused regression test and lintrunner -a both pass locally; CI has restarted on the refreshed head.
